packages/gatsby/src/query/query-compiler.js

```diff
@@ -1,384 +1,400 @@
-              start: locInGraphQlToLocInFile(
-                doc.definitions[0].templateLoc,
-                error.locations[0]
-              ),
-            }
-            return errorParser({ message: error.message, filePath, location })
-          })
-        )
+  return processDefinitions({
+    schema,
+    operations,
+    definitionsByName,
+    addError,
+    parentSpan,
+  })
+}
 
-        boundActionCreators.queryExtractionGraphQLError({
-          componentPath: filePath,
-        })
-        return compiledNodes
-      }
+const preValidationRules = [
+  LoneAnonymousOperationRule,
+  KnownTypeNamesRule,
+  FragmentsOnCompositeTypesRule,
+  VariablesAreInputTypesRule,
+  ScalarLeafsRule,
+  PossibleFragmentSpreadsRule,
+  ValuesOfCorrectTypeRule,
+  VariablesInAllowedPositionRule,
+]
 
-      // The way we currently export fragments requires duplicated ones
-      // to be filtered out since there is a global Fragment namespace
-      // We maintain a top level fragment Map to keep track of all definitions
-      // of the fragment type and to filter them out if theythey've already been
-      // declared before
-      doc.definitions = doc.definitions.filter(definition => {
-        if (definition.kind === Kind.FRAGMENT_DEFINITION) {
-          const fragmentName = definition.name.value
-          if (fragmentMap.has(fragmentName)) {
-            if (print(definition) === fragmentMap.get(fragmentName)) {
-              return false
-            }
-          } else {
-            fragmentMap.set(fragmentName, print(definition))
+const extractOperations = (schema, parsedQueries, addError, parentSpan) => {
+  const definitionsByName = new Map()
+  const operations = []
+
+  for (const [filePath, doc] of parsedQueries.entries()) {
+    const errors = validate(schema, doc, preValidationRules)
+
+    if (errors && errors.length) {
+      addError(
+        ...errors.map(error => {
+          const location = {
+            start: locInGraphQlToLocInFile(
+              doc.definitions[0].templateLoc,
+              error.locations[0]
+            ),
           }
-        }
-        return true
-      })
+          return errorParser({ message: error.message, filePath, location })
+        })
+      )
 
-      documents.push(doc)
-      doc.definitions.forEach((def: any) => {
-        const name: string = def.name.value
-        namePathMap.set(name, filePath)
-        nameDefMap.set(name, def)
+      actions.queryExtractionGraphQLError({
+        componentPath: filePath,
       })
+      continue
     }
 
-    let compilerContext = new GraphQLCompilerContext(this.schema)
-    try {
-      compilerContext = compilerContext.addAll(
-        ASTConvert.convertASTDocuments(
-          this.schema,
-          documents,
-          validationRules,
-          RelayParser.transform.bind(RelayParser)
+    doc.definitions.forEach((def: any) => {
+      const name: string = def.name.value
+      const text = print(def)
```

Comment:
Should we move this inside the `if (def.kind === Kind.FRAGMENT_DEFINITION) ` as we only want to do it while needed? (unsure how expensive this is)